### PR TITLE
Update library description

### PR DESCRIPTION
**Pull request by hasenradball**
# Update function descriptions
- add function documentation 
# update Doxygenfile
- add Doxygen file for docu build
# update Latex Documentation
- add files after latex build

Hi Rob,
I am not sure if is worth to keep all the files from latex build.

### CODE REVIEW

**Comment by RobTillaart** [comment]
Please remove the output of doxygen (HTML + LATEX), It takes unneeded time to review / click. 
Just want to know that it can be generated, and if not it should give a failed build.


Did some homework last night and created this workflow (not 100%), it should build the documentation.
Still needs success / fail flag and clean up the generated output.

to be added under .github/workflows

doxygen.yml
```
name: DOXYGEN

on: [push, pull_request]
jobs:
  doxygen:
    runs-on: ubuntu-latest
    timeout-minutes: 10
    steps:
      - uses: actions/checkout@v4
      - uses: mattnotmitt/doxygen-action@v1.9.8
        with:
          working-directory: '.'
          doxyfile-path: 'doxyfile'
      - run: |
          ls -alR
```

**Comment by RobTillaart** [comment]
I missed the PDF in the 100+ files, 
As it is only one file, yes, keeping one PDF doesn't clutter my disks 

Propose to put it the doxygen folder instead of in latex

**Comment by RobTillaart** [comment]
Now I have 100+ new files to review, that just doesn't work for me.

**Comment by RobTillaart** [comment]
Can you add that doxygen workflow I proposed above to see if and how it works?

**Comment by RobTillaart** [comment]
> what is doxyfile path?

The path to the doxyfile starting from the base of the checked out repo. 

See - https://github.com/mattnotmitt/doxygen-action

**Comment by RobTillaart** [comment]
@hasenradball 
Thanks for all your work,

The doxygen action ran successfully (don't know if it can fail yet)
- the files are generated based upon the doxyfile, so people can create their HTML TEX whatever.
- ls -AlR part gives way to much output ==> should become ls -alR doxigen

In short, I am quite happy so far, thanks again

To be continued later (want to finish some other stuff today too)

**Comment by RobTillaart** [comment]
@hasenradball 

Q: why has the .cop a copy of all doxy comments?
This is doing work twice, which is very inefficient and error prone from maintenance point if view.

**Comment by RobTillaart** [comment]
@hasenradball 

Merge conflict comes from merging your other PRs.

**Comment by hasenradball** [comment]
Good morning Rob,

thank you for merging my PRs and for the confidence.
I will check the conflicts and greate a latest pdf, then.
What do you mesn with the cop file was not aware.

best Frank

PS: when merging the confllicts should I increase the version number?

**Comment by hasenradball** [comment]
If you don‘t mind I see similar improvements for the MCP23008.
😊

**Comment by RobTillaart** [comment]
> What do you mesn with the cop file was not aware.

I meant the PCF8574.cpp file.

It is sufficient to create the doxy comments only in the PCF8574.h file.
All functions and variables are defined in the header file.

**Comment by RobTillaart** [comment]
> What do you mesn with the cop file was not aware.

I meant the PCF8574.cpp file.

It is sufficient to create the doxy comments only in the PCF8574.h file.
All functions and variables are defined in the header file.

**Comment by RobTillaart** [comment]
>  when merging the confllicts should I increase the version number?

Can do, should be done eventually anyway.

As there is no functional change in the code the version should become 0.4.2

**Comment by hasenradball** [comment]
This fucking doxygen stuff is a great shit :1st_place_medal:

**Comment by RobTillaart** [comment]
Please add a link to the doxygen website in the readme.md in the doxygen section.
- https://www.doxygen.nl/manual/index.html

Optionally use the link of the installation page

I have a busy day ahead, so probably only time tonight.

**Comment by hasenradball** [comment]
whats your intention?
can you please make a suggestion.
Thanks

**Comment by RobTillaart** [comment]
very little time - other work has prio - so my comments come as snippets. 

**.gitignore** should not be in a repository as the contents depends on the development system the user uses.
Yours seems to be Visual Studio, but there are many others and the **gitignore** would cause overwrites or other types of conflict.
So please remove from the PR.

BTW, the filename **.giignore** is incorrect.

**Comment by RobTillaart** [comment]
> whats your intention?

TO have only doxy comments in the .h file as there the whole API for the user exists.
Duplicate them is a waste of resources.

**Comment by hasenradball** [comment]
the point is. 
I introduced the gitignore mainly to keep the files generated from doxy out of scope.
This was the main reason.
Sorry for the typo.

Remove it anyway or keep and add only the doxy files which not wanted?

**Comment by hasenradball** [comment]
would suggest to keep the gitignore file only with the doxyfiles.

**Comment by hasenradball** [comment]
> comments in the .h

Hi Rob,
yeah the best way is to have the whole comments in header file, wen can do that.
No problem, but be aware that you don' t see the whole function deklaration in one shot.

But normally the best is to have the doxy comments in header.
And then keep the cpp file clean.
Should we do so?
And waht about the File description?
Also have it in one file?
Or keep it in both.

I would also think about the version number. It is 3 times defined (2 times in Header one time in cpp)

Please give me a sign what to do:
a) with doxy comments
b) with gitignore

Best frank

**Comment by RobTillaart** [comment]
I call it a day, have been working for almost 15 hours today,
Tomorrow will be a busy day again.

**Comment by hasenradball** [comment]
Hi Rob,

did an update now.

please have a look updated the ignore file, would suggest to keep it like this, otherwise you will have 100 or more files in your git area to stage. It is better keep it like this.
- updated gitignore file
- removed all documentation for function in cpp file => copied to header file
- files looks clean and better now :-)
- updated pdf

**Comment by hasenradball** [comment]
I am not 100% sure about the file description, but it is sufficient to keep Filedescription in *.h and remove it from *.cpp file.

**Comment by hasenradball** [comment]
So take a look I think it is very nice now.
My own libs don't look so perfekt :-)

**Comment by RobTillaart** [comment]
Good morning,

Insight of last night is to put the .gitignore file itself in .gitignore.

That allows you to have it as you need, no clutter with 100+ files staged, and every other user can have its own version (or none) without disruptive interference.

Opinion?

**Comment by hasenradball** [comment]
Good morning,

did not finally catch what you mean by this:
`gitignore file itself in .gitignore.`

**Comment by RobTillaart** [comment]
If you put the filename .gitignore into .gitignore, it will be used by the git commands, but not appear in e.g. **git status**

**Comment by hasenradball** [comment]
ah understood what you mean.

putting itself gitignore file into .gitignore.
Never did before but ok for me.

**Comment by hasenradball** [comment]
Added .gitignore to the file itself.
But I think it will not work.
What was your intention with this?

**Comment by hasenradball** [comment]
then I did the correct way.

**Comment by RobTillaart** [comment]
You still have to do ```git rm .gitignore``` as it was committed earlier

**Comment by hasenradball** [comment]
Sorry did not catch wat you mean.
If I do so , `git rm .gitignore` then it will be deleted from the repo.

**Comment by hasenradball** [comment]
Sorry for the misunderstanding.

File: PCF8574.cpp [hunk 1]
```diff
@@ -1,29 +1,16 @@
...
-//    FILE: PCF8574.cpp
-//  AUTHOR: Rob Tillaart
-//    DATE: 02-febr-2013
-// VERSION: 0.4.1
-// PURPOSE: Arduino library for PCF8574 - 8 channel I2C IO expander
-//     URL: https://github.com/RobTillaart/PCF8574
-//          http://forum.arduino.cc/index.php?topic=184800
-
-
 #include "PCF8574.h"
```
Comment (RobTillaart) [inline PCF8574.cpp:1]: Lost the header of the cpp file

**Comment by hasenradball** [comment]
No you mentioned in the Discussion it brings no benefit to keep.
So please make clear statements.

File: PCF8574.cpp (hunk 1 above)
Comment (hasenradball) [reply PCF8574.cpp:1]: image

**Comment by RobTillaart** [comment]
I had a quick review, hope to find more time coming weekend, but at the moment very busy.

most comments are rather small, I think you did a very good job sofar!

**Comment by RobTillaart** [comment]
> No you mentioned in the Discussion it brings no benefit to keep.
> So please make clear statements.

Sorry my mistake to not being clear enough.

**Comment by hasenradball** [comment]
Hi,

any news for this PR?

**Comment by hasenradball** [comment]
Hello Rob,

could you have a look to this PR?

**Comment by RobTillaart** [comment]
@hasenradball 

I'm ill and am not able to work. 
All my activities stopped.
Be aware this might take some weeks. 
Sorry for the inconveniences,

**Comment by hasenradball** [comment]
Hi Rob,

oh my god sorry for that.
Wish you soon recovery.
Will wait for your feedback then.

one point I had in mind with docxygen you will be able with github pages to update the github page.
Do you have expirience in that point?

szenario would be that if main branch has a push a doxy run will update the github page.

**Comment by hasenradball** [comment]
Here you see an example of automated Documentation creation.

[AM2302-Sensor](https://hasenradball.github.io/AM2302-Sensor/)

**Comment by RobTillaart** [comment]
@hasenradball 
Not recovered yet, will work a few hours this week.
Focus is on my work so my Arduino activities are on a minimal level.
Sorry for the inconveniences,

**Comment by hasenradball** [comment]
Hi Rob,

recover soon and have a nice christmas

**Pull request closed without merging**